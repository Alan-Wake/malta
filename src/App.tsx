import * as React from "react";
import { ThemeProvider } from "styled-components";
import theme from "./theme";

export class App extends React.Component<any, any> {
  constructor(props: any) {
    super(props);
  }

  public render() {
    return (
      <ThemeProvider theme={theme}>
        {this.props.children}
      </ThemeProvider>
    );
  }
}
