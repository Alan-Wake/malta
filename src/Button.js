import * as React from 'react';

import classNames from 'classnames';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const propTypes = {
    className: PropTypes.string
};

const defaultProps = {
    tag: "div"
};

const StyledButton = styled.button`
  padding: 1rem;
  font-size: 1rem;
  font-weight: bold;
  cursor: pointer;
  width: 100%;
  color: ${variable.blue};
  background: #22409a;
  &:hover {
    background: #425FF3;
    color: #ffffff;
  }
`;

const Button = (props) => {
    const {
        className,
        ...attributes
    } = props;

    const classes = classNames("sb-btn", className);
    return (
        <StyledButton className={classes}>테스트</StyledButton>
    );
}

Button.propTypes = propTypes;
Button.defaultProps = defaultProps;

export default Button;