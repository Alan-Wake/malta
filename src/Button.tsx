import * as React from "react";
import styled from "styled-components";
const {
  borderColor,
  borderRadius,
  borders,
  color,
  fontSize,
  height,
  space,
  width,
} = require("styled-system");

const StyledButton = styled.button`
  align-items: center;
  cursor: pointer;
  display: flex;
  justify-content: center;
  ${borderColor}
  ${borderRadius}
  ${borders}
  ${color}
  ${fontSize}
  ${space}
  ${width}
  ${height}
`;

export interface IProps {
  onClick?: React.MouseEventHandler<HTMLElement>;
  type?: string;
  // borderColor
  borderColor?: string;
  // borderRadius
  borderRadius?: number | string;
  // borders
  border?: string;
  borderTop?: string;
  borderRight?: string;
  borderBottom?: string;
  borderLeft?: string;
  // color
  color?: string;
  bg?: string;
  // fontSize
  fontSize?: number | string;
  // height
  height?: number | string;
  // space
  m?: number | string;
  mt?: number | string;
  mr?: number | string;
  mb?: number | string;
  ml?: number | string;
  mx?: number | string;
  my?: number | string;
  p?: number | string;
  pt?: number | string;
  pr?: number | string;
  pb?: number | string;
  pl?: number | string;
  px?: number | string;
  py?: number | string;
  // width
  width?: number | string;
}

export class Button extends React.Component<IProps, any> {
  public static defaultProps: IProps = {
    onClick: () => {},
    // borderColor
    borderColor: "transparent",
    // borderRadius
    borderRadius: "30px",
    // borders
    border: "none",
    // color
    color: "white",
    bg: "blue.royal",
    // fontSize
    fontSize: "h3",
    // height
    height: "60px",
    // space
    m: 0,
    mb: 2,
    // width
    width: "auto",
  };

  public render() {
    return (
      <StyledButton
        {...this.props}
      />
    );
  }
}
