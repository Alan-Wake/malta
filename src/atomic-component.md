# Malta Component Design

현재 malta를 구성하는 컴포넌트는,

```
- core
- salesboost
```

되어있는것을 보았으며 각 코드를 구성하는 방식을 읽어보았습니다.

모든 컴포넌트를 구현하고, 디자인 철학을 적용 할수 없지만,디자인 시스템을 구축하는 방식에 대하여 고찰이 필요하다고 생각이 듭니다. 컴포넌트를 구성하는 방식은 여러개가 있을것 같으나, 제가 경험한 재사용성 이 높은 디자인에 대해서 작성 해보려 합니다.

---

## Atomic Component Design

atomic component design 에 대해서 자세하게 더 알고자 하신다면 아래의 참고링크를 통해서 할 수 있습니다.

[참고 링크](http://atomicdesign.bradfrost.com/chapter-2/)

atomic component design은 많은 React Component를 다루는 [예시](https://github.com/diegohaz/arc)들이 많은것 같습니다.

atomic component design을 도입하는데 있어서
가장 중요한것은 **"정답은 없다."** 입니다.

그 이유는 아래와 같습니다.

- 디자인 팀에서 생각하는 Button 과 개발자들이 생각하는 Button 이 다를 수 있습니다.

- 개발자 사이에서도 atom을 나누는 기준에 대해서 서로 대립 관계가 형성 될수 있습니다.

그렇기에 component를 정의하는데 있어서 디자인팀에서 명시 되어 있는 팀내 component design guide를 통해
디자인팀과 컴포넌트화에 대해 1차적으로 논의가 되어야하며,

2차적으로 개발자들과의 커뮤니케이션을 통해서 팀내 가이드를 정하는것이 맞는것 같습니다.

---

### 어떻게 진행 되어야 할까?

개발자내에서 가장 많이 빈번하게 발생하는 경우는 2가지 입니다.

#### 예시 1)atom + atom = molecule 이라는 공식.

: Input Component를 atom component로 만들때,
label, error message 가 포함 되어야 하는것일까?
이미 존재 하는 Text Atom을 Input에 넣게되는게 맞을까?

라는 것에 대해서 팀내에서 맞다 아니다를 정하게 되면,
쉽게 해결된 문제 이지만 이것 또한 팀 가이드를 명확하게 정의 해야 합니다.

#### 예시 2) 컴포넌트가 추가 할때마다 유사한 컴포넌트가 생긴다.

경험담이지만, 많은 개발자가 관리를 하게 되면서 atom을 가볍게 만들게 되는데, 현재 "/src/salesboost"에 있는 IconButton 이 해당 되는 경우 라고 판단이 됩니다. Icon, Button 이라는 존재하는 Atom들에서 새로운 컴포넌트가 생성 되었을때 Atom Component를 정의 내리는 기준은

_"These atoms include basic HTML elements like form labels, inputs, buttons, and others that can’t be broken down any further without ceasing to be functional."_

으로 쉽게 보자면 더이상 쪼개기 힘들고, 한가지의 기능을 가진 기본적인 element라고 정의 내릴수있습니다.

여기서 functional 에 집중하자면, Icon은 하나의 이미지 이며, Button 은 클릭이 가능한 것 이라고 정의를 내릴수 있을것 같습니다.

초기에는 Design, Function 으로 기준을 나누기 모호할때는 기능에 집중하는것이 더 명쾌해 질수 있을것같습니다.

예시)

```js
render() {
  <>
    <Button type="icon" name="arrow" />
    <Button label="Text" />
  </>
}
```

---

#### 예시 3) molecule 관리 방식

실제 개발을 시작하게 되면, malta project 와 실제 service project를 동시에 진행 하다보면, malta project에는 molecule을 다양하게 만드는 경우가 많이 발생하였습니다. 쉽게 예시를 들자면 LabelWithInput 과 InputWithLabel처럼 같은 기능을 하는것이 중복 되는경우가 발생이 됩니다.

이러한 것을 방지하기 위해서는 아래의 방식이 필요합니다.

- 팀의 molecule component 정의
- naming convention

**molecule component?**

_1.molecules are relatively simple groups of UI elements functioning together as a unit._

_2.Creating simple components helps UI designers and developers adhere to the single responsibility principle, an age-old computer science precept that encourages a “do one thing and do it well” mentality._

1개의 기능이 아닌 1가지 성질을 지니고 잇으며, 1가지 성질 만을 하는것을 단일 원칙으로 만드는것을 molecule component 으로 생각하면 될 것 같습니다.

- atom + atom = molecule
- atom + moleclue = molecule

**예시1) naming에는 구성된 컴포넌트 보다는 기능(성질)에 중점을 준다.**

실제로 Refactor를 진행하기전에 molecule folder에 있는 컴포넌트들을 나열 하다보면, 디자인에 따라서 혹은 컴포넌트들의 관계에 따라서 labelWithInput, InputWithLabel 처럼 같은 기능을 가지고 있지만 이름에 따라서 중복되는 경우가 많이 발생합니다.

그렇기에 naming을 할 때는 서비스 에서 쓰이는 UI unit을 구현함을 목표로 하면 될 것 같습니다.

예시)

```js
render() {
  <>
    <Progress type="spinner"/>
    <Progress type="linear" />
    <Progress type="dash-board" />
  <>
}
```

---

#### 예시 4) 실제 프로젝트에선?

실제 프로젝트에서는 비지니스 로직에 집중하게 되면, 공용 컴포넌트 프로젝트를 자주 관리하지 못하는 경우가 많이 생기거나, 비지니스 로직에서 유사 컴포넌트를 직접 제작하는 경우가 발생합니다. 이럴때를 위해서 개발팀 내에서 이슈 발생시 한가지 법칙을 만들면 좋을것 같습니다

**유사하다 판단되는것이 3번이 되면 Refactor하자!**
