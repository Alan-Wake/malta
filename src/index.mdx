---
name: About
route: /
---

# Malta [[Github]](https://github.com/salesboostio/malta)

Salesboost UI library. Atomic Design 패러다임을 적용해 컴포넌트들이 추상화의
수준에 맞춰 카테고리가 나뉘어져 있습니다. 카테고리의 hierarchy는 `atom` - `molecule` - `organism` - `template`
순이며 모든 컴포넌트는 자신보다 상위레벨의 컴포넌트를 참조해선 안됩니다.

- [Atom](./src-atoms-index): Text, Box, Image...
- [Molecules](./src-molecules-index): Shadow, Label, Button, Span, HyperLink, Chip...
- [Organisms](./src-organisms-index): Card, FixedHeader, FeedItem, GridItem, Slider...
- [Templates](./src-templates-index): SalesBoostFooter, SalesBoostHeader, CatalogNav...
- [Misc](./src-misc-index): 단독으로 사용되지 않고 다른 컴포넌트들을 꾸며줄 수 있는 유틸리티들 및, 기타 분류되지 않은 컴포넌트들을 포함합니다.
- [Commons](./src-commons-index): 공통으로 사용하는 App, Theme

## Development Workflow
malta 개발 워크플로우에 대한 설명입니다. github 사용법에 대해 이미 알고 있는 경우 4번부터 보면 됩니다.

1. [github.com/salesboostio/malta](https://github.com/salesboostio/malta)에서 레포지토리를 fork합니다.
2. `git clone git@github.com:<YOUR_ID>/malta.git`
3. `git remote add upstream git@github.com:salesboostio/malta.git`
4. `git checkout atomic && git checkout -b "<your_new_branch>"`
5. `npm install && npm run docz` 명령어로 docz 서버를 실행합니다. [localhost:3000](http://localhost:3000)
6. `npm run new` 명령어로 새로운 컴포넌트를 생성합니다. [[Tip참고]](./tips)
7. 편집이 완료되었으면 `npm run build && npm run lint && npm run test`로 테스트 합니다.
8. `git add * && git commit -m "<commit_message> && git push origin "<your_new_branch>"`로 본인의 깃에 올린뒤 pull request를 생성합니다. 

## Usage Workflow
malta를 사용하는 방법에 대한 설명입니다. 프로젝트 내에서 몰타를 사용할 때 참고하면 됩니다.

1. package.json의 dependencies에  `"malta": "git+ssh://git@github.com/salesboostio/malta.git#atomic"` 를 추가합니다. 
   salesboostio/frontend-boilerplate를 사용하고 있는 경우 이미 추가가 되어 있습니다.
2. 소스코드 내에서 `import { Box, Flex, ... } from "malta";` 로 컴포넌트들을 사용할 수 있습니다.
3. 컴포넌트들을 확장하고싶을경우 devDependencies에 `styled-system`, `styled-components`를 추가해줍니다. 
4. const ExtendedBox = Box.extend`...`; 로 확장해서 사용할 수 있습니다.

## Shell Command
```
~/malta$ tsc             # 컴파일
~/malta$ npm run new     # 새 컴포넌트 생성 cli   
```

## Docz [[link]](https://www.docz.site/)
프로젝트 다큐멘테이션 라이브러리며 mdx를 사용합니다. 소스 폴더내에 어디든 .mdx 파일을 생성하기만 하면 
자동으로 docz 웹페이지로 변환시켜줍니다. 
```
npm run docz
npm run docz build
```


## References