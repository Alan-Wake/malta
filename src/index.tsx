// require('./bootstrap-utilities.min.css');

// import * as React from 'react';
// import * as ReactDOM from 'react-dom';

export { default as Button } from './button';
export { Row, Col } from './grid';
export { default as Input } from './input';
export { App, default as Layout } from './layout';