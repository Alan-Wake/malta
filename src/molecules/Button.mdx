---
name: Button
menu: Molecules
---

import { Playground } from "docz"
import { App, Chip, Flex, Button } from "../../dist"

# Button

This is Button

## Dependencies
<App>
<Flex>
<Chip bg="blue.deep" color="#FFF">None</Chip>
</Flex>
</App>

## Properties

|Property|Type|Default|
|-|-|-|
| color | string | white.standard |
| height | ResponsiveValues | [52px, 60px] |
| width | ResponsiveValue | fit-content |
| fontSize | ResponsiveValues | [h5, h3] |
| hoverColor | String | color.blue.deep |
| focusColor | String | color.blue.dark |
| bg | String | blue.royal |
| onClick | function | ()=>{} |

- hoverColor, focusColor는 theme적용이 안되므로 theme을 import해서 color.*으로 사용했습니다.

## Usecase

<Playground>
  <App>
    <Button>Click me</Button>
  </App>
</Playground>

## Best Practices

## Test