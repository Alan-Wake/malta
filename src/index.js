require('./bootstrap-utilities.min.css');

import Button from './Button';
import Navbar from './Navbar';

export { 
    Button,
    Navbar 
};