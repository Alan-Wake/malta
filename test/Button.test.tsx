
import * as React from "react";
import { create } from "react-test-renderer";
import { Button } from "../src/Button";

test("Say my name, say my name...", () => {
    const tree1 = create(
        <Button>Hello World!</Button>
    ).toJSON();
    expect(tree1).toMatchSnapshot();
});
