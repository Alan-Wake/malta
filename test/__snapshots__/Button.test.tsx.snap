// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Say my name, say my name... 1`] = `
<button
  className="sb-button sc-bdVaJa kKJYgi"
>
  Hello World!
</button>
`;
